ocs_ci/ocs/external_ceph.py

```diff
@@ -0,0 +1,904 @@
+        eg: self.exec_cmd(cmd='uptime')
+        or
+        self.exec_cmd(cmd='background_cmd', check_ec=False)
+
+        Attributes:
+            check_ec (bool): False will run the command and not wait for exit
+                code
+
+        """
+
+        if kw.get('sudo'):
+            ssh = self.rssh
+        else:
+            ssh = self.ssh
+
+        if kw.get('timeout'):
+            timeout = kw['timeout']
+        else:
+            timeout = 120
+        logger.info("Running command %s on %s", kw['cmd'], self.ip_address)
+        stdin = None
+        stdout = None
+        stderr = None
+        if self.run_once:
+            self.ssh_transport().set_keepalive(15)
+            self.rssh_transport().set_keepalive(15)
+        if kw.get('long_running'):
+            logger.info("long running command --")
+            channel = ssh().get_transport().open_session()
+            channel.exec_command(kw['cmd'])
+            read = ''
+            while True:
+                if channel.exit_status_ready():
+                    ec = channel.recv_exit_status()
+                    break
+                rl, wl, xl = select([channel], [], [channel], 4200)
+                if len(rl) > 0:
+                    data = channel.recv(1024)
+                    read += data.decode()
+                    logger.info(data.decode())
+                if len(xl) > 0:
+                    data = channel.recv(1024)
+                    read += data.decode()
+                    logger.info(data.decode())
+            return read, ec
+        try:
+            stdin, stdout, stderr = ssh().exec_command(
+                kw['cmd'], timeout=timeout)
+        except SSHException as e:
+            logger.error("Exception during cmd %s", str(e))
+            if 'Timeout openning channel' in str(e):
+                logger.error("channel reset error")
+        exit_status = stdout.channel.recv_exit_status()
+        self.exit_status = exit_status
+        if kw.get('check_ec', True):
+            if exit_status == 0:
+                logger.info("Command completed successfully")
+            else:
+                logger.error("Error during cmd %s, timeout %d", exit_status, timeout)
+                raise CommandFailed(kw['cmd'] + " Error:  " + str(stderr.read().decode()) + ' ' + str(self.ip_address))
+            return stdout, stderr
+        else:
+            return stdout, stderr
+
+    def write_file(self, **kw):
+        if kw.get('sudo'):
+            client = self.rssh
+        else:
+            client = self.ssh
+        file_name = kw['file_name']
+        file_mode = kw['file_mode']
+        ftp = client().open_sftp()
+        remote_file = ftp.file(file_name, file_mode, -1)
+        return remote_file
+
+    def _keep_alive(self):
+        while True:
+            self.exec_command(cmd='uptime', check_ec=False)
+            sleep(60)
+
+    def reconnect(self):
+        pass
+
+    def __getstate__(self):
+        d = dict(self.__dict__)
```

Comment:
`d` doesn't describe what this dict represents. Can we come up with a better name?

Reply:
done